lua/telescope/builtin/files.lua

```diff
@@ -186,19 +186,24 @@ files.current_buffer_fuzzy_find = function(opts)
 end
 
 files.tags = function(opts)
-  local ctags_file = opts.ctags_file or 'tags'
-
-  if not vim.loop.fs_open(vim.fn.expand(ctags_file), "r", 438) then
-    print('Tags file does not exists. Create one with ctags -R')
-    return
+  local tagfiles
+  if opts.ctags_file then
+    tagfiles = {opts.ctags_file}
+  else
+    tagfiles = vim.fn.map(vim.fn.tagfiles(), 'fnamemodify(v:val, ":p")')
+    if vim.tbl_isempty(tagfiles) then
+      print('No tags file found. Create one with ctags -R or verify your &tagfiles option is correctly set.')
+      return
+    end
   end
 
-  local fd = assert(vim.loop.fs_open(vim.fn.expand(ctags_file), "r", 438))
-  local stat = assert(vim.loop.fs_fstat(fd))
-  local data = assert(vim.loop.fs_read(fd, stat.size, 0))
-  assert(vim.loop.fs_close(fd))
-
-  local results = vim.split(data, '\n')
+  local results = {}
+  for _, each in ipairs(tagfiles) do
+    tags_directory = vim.fn.fnamemodify(each, ":h")
+    for line in io.lines(each) do
```

Comment:
Could we not use `io.lines`? Its super slow compared to `vim.loop.fs_`
I was think about either this
```lua
local fd = vim.loop.fs_open(path, "r", 438)
local stat = vim.loop.fs_fstat(fd)
for i = 0, stat.size do -- might be a off by one issue. Not sure right now. Writing here in github
  local line = assert(vim.loop.fs_read(fd, 1, i))
end
vim.loop.fs_close(fd)
```
Or we could do it async with finder.new_oneshot_job like in #274. As command you could do something like this:
`flatten { 'grep', "", tagfiles }` that would give use 'filename:line' in `entry_maker` as input. So you might have to change the regex to get the tagfile_name. 
Benefit: Picker would be there asap and lines will come after one by one.

Reply:
I figured there was a reason the code didn't use `io.lines` already but that someone'd tell me what it was if I used it :).

Will have a look at #274.

Reply:
What's the crazy magic 438 that appears all over the place?

Reply:
It's the mode. There are also some `luv.constants` which can be used for that. But i don't know i just used what they documented.
Take a look at [luv docs](https://github.com/luvit/luv/blob/master/docs.md#file-system-operations) which is `vim.loop` for more information. I am about to go to sleep. So good luck and have fun fixing my all my nit pickings :)
Also thanks for the PR. I didn't know about `&tagfiles` when i implemented this. I basically just took fzf as inspiration. So i appreciate you fixing my stuff.

Reply:
Oh right decimal-octal, fun.

And yeah all good! Thanks for the comments.

Reply:
Oh i just realized. When you do it with `grep` you should ues `-H` option, so the filename will always be displayed. Even for just one file.

Reply:
I can't seem to get this working with finder.new_oneshot_job.

I get yelled at for calling vim functions in a libuv callback.

It seems part of #274 fixed that but even with the fix there I get the same warning.

It also seems a bit silly to me to not have a way to read files from Lua. I was able to just straight up copy a function from luvit to do so -- will push that in a bit to get your thoughts, if there was a way to just populate a finder incrementally in a callback I think this'd work. I thought that'd be `finder.new_job` but maybe I'm not using it right.

Reply:
The thing is we can't use any vimL inside a `coroutine` but if we would do a `new_oneshot_job` we would use a `coroutine`.
Inside `entry_display.lua` (which is used to dynamically generate tables) we also truncate and to determine the string display width we use `vim.fn.strdisplaywidth(str)` which is vimL. Tj thought maybe `vim.api.nvim_strwidth` would help but it did not. So right now we could use `#str` to determine the string size but that does not work for any charakter that requires two cells(emojis or chinese characters i think).

I will take a look at this tomorrow. Its already to late for me :smile:.

I will open new converstation if there is something to talk about so you can close this one for now. Thanks for your work